File: Source/MKL/examples/vmlc/source/vssincos.c
#include <stdio.h>
#include <math.h>
#include "func_interv.h"

#include "mkl_vml.h"

int main()
{
  float fA1[VEC_LEN],fB1[VEC_LEN],fB2[VEC_LEN],fB3[VEC_LEN],fB4[VEC_LEN];

  int i=0,vec_len=VEC_LEN;
  double CurRMS,MaxRMSsin=0,MaxRMScos=0;

  for(i=0;i<vec_len;i++) {
    fA1[i]=(float)(__VML_SSINCOS_BEG+((__VML_SSINCOS_END-__VML_SSINCOS_BEG)*i)/vec_len);
    fB1[i]=0;
    fB2[i]=0;
    fB3[i]=(float)sin(fA1[i]);
    fB4[i]=(float)cos(fA1[i]);
  }

#if defined(MKL_VML_MIXED)
  vsSinCos(vec_len,fA1,fB1,fB2);
#elif defined(MKL_VML_LOWER)
  vssincos(&vec_len,fA1,fB1,fB2);
#elif defined(MKL_VML_UPPER)
  VSSINCOS(&vec_len,fA1,fB1,fB2);
#else
# error You must define MKL_VML_CDECL for CDECL call convention or MKL_VML_STDCALL for STDCALL call convention.
#endif


  printf("vsSinCos test/example programm\n\n");
  printf("        Argument         vsSinCos:   Sin        Cos           Sin        Cos\n");
  printf("===============================================================================\n");
  for(i=0;i<vec_len;i++) {
    printf("% 22.14f % 20.6f % 10.6f % 13.6f % 10.6f\n",fA1[i],fB1[i],fB2[i],fB3[i],fB4[i]);
    CurRMS=(fB1[i]-fB3[i])/(0.5*(fB1[i]+fB3[i]));
    if(MaxRMSsin<CurRMS) MaxRMSsin=CurRMS;
    CurRMS=(fB2[i]-fB4[i])/(0.5*(fB2[i]+fB4[i]));
    if(MaxRMScos<CurRMS) MaxRMScos=CurRMS;
  }
  printf("\n");
  printf("Maximum relative error (sin): %.2f\n",MaxRMSsin);
  printf("Maximum relative error (cos): %.2f\n",MaxRMScos);
}

File: Source/MKL/examples/vmlf/source/vssincos.f
	programm MKL_VML_TEST

	include "func_interv.fi"

	real*4 sA(VEC_LEN)
	real*4 sB1(VEC_LEN)
	real*4 sB2(VEC_LEN)
	real*4 sB3(VEC_LEN)
	real*4 sB4(VEC_LEN)
	real*4 CurRMS1,MaxRMS1
	real*4 CurRMS2,MaxRMS2
	real*4 MaxRMS
	integer i

	MaxRMS=0.0
	do i=1,VEC_LEN
		sA(i)=VML_SSINCOS_BEG+((VML_SSINCOS_END-VML_SSINCOS_BEG)*i)/VEC_LEN
		sB1(i)=0;
		sB2(i)=0;
		sB3(i)=sin(sA(i))
		sB4(i)=cos(sA(i))
	end do

	call VSSINCOS(VEC_LEN,sA,sB1,sB2)

	print *,"vsSinCos test/example programm"
	print *,""
	print *,""
	print *,"           Argument                     vsSinCos                      SinCos"
	print *,"=============================================================================="
	do i=1,VEC_LEN
		print 10,sA(i),sB1(i),sB2(i),sB3(i),sB4(i)
		CurRMS1=(sB1(i)-sB3(i))/(0.5*(sB1(i)+sB3(i)))
		CurRMS2=(sB2(i)-sB4(i))/(0.5*(sB2(i)+sB4(i)))
		if(MaxRMS1<CurRMS1) MaxRMS1=CurRMS1
		if(MaxRMS2<CurRMS2) MaxRMS2=CurRMS2
	end do

	if(MaxRMS2<MaxRMS1) MaxRMS=MaxRMS1
	if(MaxRMS2>MaxRMS1) MaxRMS=MaxRMS2
	print *,""
	print 11,"Maximum relative error: ",MaxRMS

10	format(F22.14,F20.6,F10.6,F13.6,F10.6)
11	format(A,F25.14)

	end
